server/rating.py

```diff
@@ -2,58 +2,126 @@
 Type definitions for player ratings
 """
 
-from typing import DefaultDict, Optional, Tuple, TypeVar, Union
+from dataclasses import dataclass
+from typing import Dict, Optional, Set, Tuple, TypeVar, Union
 
-from trueskill import Rating
+import trueskill
 
+from server.config import config
+from server.weakattr import WeakAttribute
 
-# Values correspond to legacy table names. This will be fixed when db gets
-# migrated.
+Rating = Tuple[float, float]
+V = TypeVar("V")
+
+
+@dataclass(init=False)
+class Leaderboard():
+    id: int
+    technical_name: str
+    initializer = WeakAttribute["Leaderboard"]()
+
+    def __init__(
+        self,
+        id: int,
+        technical_name: str,
+        initializer: Optional["Leaderboard"] = None
+    ):
+        self.id = id
+        self.technical_name = technical_name
+        if initializer:
+            self.initializer = initializer
+
+    def __repr__(self) -> str:
+        initializer = self.initializer
+        initializer_name = "None"
+        if initializer:
+            initializer_name = initializer.technical_name
+        return (
+            f"{self.__class__.__name__}("
+            f"id={self.id}, technical_name={self.technical_name}, "
+            f"initializer={initializer_name})"
+        )
+
+
+# Some places have references to these ratings hardcoded.
 class RatingType():
     GLOBAL = "global"
     LADDER_1V1 = "ladder_1v1"
-    TMM_2V2 = "tmm_2v2"
 
 
-K = Union[RatingType, str]
-V = TypeVar("V")
+class PlayerRatings(Dict[str, Rating]):
+    def __init__(self, leaderboards: Dict[str, Leaderboard], init: bool = True):
+        self.leaderboards = leaderboards
+        # Rating types which are present but should be recomputed.
+        self.transient: Set[str] = set()
+
+        # Initialize known rating types so the client can display them
+        if init:
+            _ = self[RatingType.GLOBAL]
+            _ = self[RatingType.LADDER_1V1]
```

Comment:
The client no longer makes any assumptions about what ratings exist and will just set them to 0 if they do not exist

Reply:
Ok cool. I'll mark this as deprecated then.